README.md
<!-- logo -->
<h1 align='center'>StarkWare Crypto Utils</h1>

<!-- tag line -->
<h4 align='center'> Signatures, keys and Pedersen hash on STARK friendly elliptic curve</h4>

<!-- primary badges -->
<p align="center">
  <a href="https://www.w3schools.com/js/">
    <img src='https://badges.aleen42.com/src/javascript.svg' />
  </a> 
  <a href="https://www.npmjs.com/package/@starkware-industries/starkware-crypto-utils">
    <img src='https://img.shields.io/npm/v/@starkware-industries/starkware-crypto-utils?label=npm' />
  </a>
  <a href="https://starkware.co/">
    <img src="https://img.shields.io/badge/powered_by-StarkWare-navy">
  </a>
</p>

## Installation

```bash
// using npm
npm i @starkware-industries/starkware-crypto-utils

// using yarn
yarn add @starkware-industries/starkware-crypto-utils
```

## How to use it

```js
const starkwareCrypto = require('@starkware-industries/starkware-crypto-utils');
```

## API

```javascript

{
    prime,
    ec: starkEc,
    constantPoints,
    shiftPoint,
    maxEcdsaVal, // Data.
    pedersen,
    getLimitOrderMsgHash,
    getTransferMsgHash,
    sign,
    verify,
    assertInRange,
    getTransferMsgHashWithFee,
    getLimitOrderMsgHashWithFee // Function.

    asset: {
      getAssetType,
      getAssetId // Function.
    },

    keyDerivation: {
      StarkExEc: ec.n, // Data.
      getPrivateKeyFromEthSignature,
      privateToStarkKey,
      getKeyPairFromPath,
      getAccountPath,
      grindKey // Function.
    },

    messageUtils: {
      assertInRange // Function.
    }
}
```

## Usage

### Signing a StarkEx order

```javascript
const starkwareCrypto = require('@starkware-libs/starkware-crypto-utils');
const testData = require('test/config/signature_test_data.json');

const privateKey = testData.meta_data.party_a_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) ===
    testData.settlement.party_a_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.settlement.party_a_order.public_key.substring(2)}`
);

const {party_a_order: partyAOrder} = testData.settlement;
const msgHash = starkwareCrypto.getLimitOrderMsgHash(
  partyAOrder.vault_id_sell, // - vault_sell (uint31)
  partyAOrder.vault_id_buy, // - vault_buy (uint31)
  partyAOrder.amount_sell, // - amount_sell (uint63 decimal str)
  partyAOrder.amount_buy, // - amount_buy (uint63 decimal str)
  partyAOrder.token_sell, // - token_sell (hex str with 0x prefix < prime)
  partyAOrder.token_buy, // - token_buy (hex str with 0x prefix < prime)
  partyAOrder.nonce, // - nonce (uint31)
  partyAOrder.expiration_timestamp // - expiration_timestamp (uint22)
);

assert(
  msgHash === testData.meta_data.party_a_order.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.party_a_order.message_hash.substring(2)
);

const msgSignature = starkwareCrypto.sign(keyPair, msgHash);
const {r, s} = msgSignature;

assert(starkwareCrypto.verify(publicKey, msgHash, msgSignature));
assert(
  r.toString(16) === partyAOrder.signature.r.substring(2),
  `Got: ${r.toString(16)}. Expected: ${partyAOrder.signature.r.substring(2)}`
);
assert(
  s.toString(16) === partyAOrder.signature.s.substring(2),
  `Got: ${s.toString(16)}. Expected: ${partyAOrder.signature.s.substring(2)}`
);

// The following is the JSON representation of an order:
console.log('Order JSON representation: ');
console.log(partyAOrder);
console.log('\n');
```

### StarkEx key serialization

```javascript
const starkwareCrypto = require('@starkware-libs/starkware-crypto-utils');

const pubXStr = publicKey.pub.getX().toString('hex');
const pubYStr = publicKey.pub.getY().toString('hex');

// Verify Deserialization.
const pubKeyDeserialized = starkwareCrypto.ec.keyFromPublic(
  {x: pubXStr, y: pubYStr},
  'hex'
);
assert(starkwareCrypto.verify(pubKeyDeserialized, msgHash, msgSignature));
```

### Signing a StarkEx order with fee

```javascript
const privateKey = testData.meta_data.party_a_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) ===
    testData.settlement.party_a_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.settlement.party_a_order.public_key.substring(2)}`
);

const {party_a_order: partyAOrder} = testData.settlement;
const feeInfo = testData.fee_info_user;
const msgHash = starkwareCrypto.getLimitOrderMsgHashWithFee(
  partyAOrder.vault_id_sell, // - vault_sell (uint64)
  partyAOrder.vault_id_buy, // - vault_buy (uint64)
  partyAOrder.amount_sell, // - amount_sell (uint63 decimal str)
  partyAOrder.amount_buy, // - amount_buy (uint63 decimal str)
  partyAOrder.token_sell, // - token_sell (hex str with 0x prefix < prime)
  partyAOrder.token_buy, // - token_buy (hex str with 0x prefix < prime)
  partyAOrder.nonce, // - nonce (uint31)
  partyAOrder.expiration_timestamp, // - expiration_timestamp (uint22)
  feeInfo.token_id, // - token (hex str with 0x prefix < prime)
  feeInfo.source_vault_id, // - fee_source_vault_id (uint31)
  feeInfo.fee_limit // - amount (uint63 decimal str)
);

assert(
  msgHash ===
    testData.meta_data.party_a_order_with_fee.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.party_a_order_with_fee.message_hash.substring(2)
);

// The following is the JSON representation of an order:
console.log('Order With Fee JSON representation: ');
// Fee info is added to the order, and will be also be seen in the JSON of Settlement.
partyAOrder.fee_info = feeInfo; // eslint-disable-line
console.log(partyAOrder);
console.log('\n');
```

### StarkEx transfer

```javascript
const starkwareCrypto = require('@starkware-libs/starkware-crypto-utils');
const testData = require('test/config/signature_test_data.json');

const privateKey = testData.meta_data.transfer_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) === testData.transfer_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
            Expected: ${testData.transfer_order.public_key.substring(2)}`
);

const transfer = testData.transfer_order;
const msgHash = starkwareCrypto.getTransferMsgHash(
  transfer.amount, // - amount (uint63 decimal str)
  transfer.nonce, // - nonce (uint31)
  transfer.sender_vault_id, // - sender_vault_id (uint31)
  transfer.token, // - token (hex str with 0x prefix < prime)
  transfer.target_vault_id, // - target_vault_id (uint31)
  transfer.target_public_key, // - target_public_key (hex str with 0x prefix < prime)
  transfer.expiration_timestamp // - expiration_timestamp (uint22)
);

assert(
  msgHash === testData.meta_data.transfer_order.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.transfer_order.message_hash.substring(2)
);

// The following is the JSON representation of a transfer:
console.log('Transfer JSON representation: ');
console.log(transfer);
console.log('\n');
```

### StarkEx conditional transfer

```javascript
const starkwareCrypto = require('@starkware-libs/starkware-crypto-utils');
const testData = require('test/config/signature_test_data.json');

const privateKey =
  testData.meta_data.conditional_transfer_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) ===
    testData.conditional_transfer_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.conditional_transfer_order.public_key.substring(
          2
        )}`
);

const transfer = testData.conditional_transfer_order;
const msgHash = starkwareCrypto.getTransferMsgHash(
  transfer.amount, // - amount (uint63 decimal str)
  transfer.nonce, // - nonce (uint31)
  transfer.sender_vault_id, // - sender_vault_id (uint31)
  transfer.token, // - token (hex str with 0x prefix < prime)
  transfer.target_vault_id, // - target_vault_id (uint31)
  transfer.target_public_key, // - target_public_key (hex str with 0x prefix < prime)
  transfer.expiration_timestamp, // - expiration_timestamp (uint22)
  transfer.condition // - condition (hex str with 0x prefix < prime)
);

assert(
  msgHash ===
    testData.meta_data.conditional_transfer_order.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.conditional_transfer_order.message_hash.substring(2)
);

// The following is the JSON representation of a transfer:
console.log('Conditional Transfer JSON representation: ');
console.log(transfer);
console.log('\n');
```

### StarkEx transfer with fee

```javascript
const privateKey = testData.meta_data.transfer_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) === testData.transfer_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.transfer_order.public_key.substring(2)}`
);

const transfer = testData.transfer_order;
const feeInfo = testData.fee_info_user;
const msgHash = starkwareCrypto.getTransferMsgHashWithFee(
  transfer.amount, // - amount (uint63 decimal str)
  transfer.nonce, // - nonce (uint31)
  transfer.sender_vault_id, // - sender_vault_id (uint64)
  transfer.token, // - token (hex str with 0x prefix < prime)
  transfer.target_vault_id, // - target_vault_id (uint64)
  transfer.target_public_key, // - target_public_key (hex str with 0x prefix < prime)
  transfer.expiration_timestamp, // - expiration_timestamp (uint22)
  feeInfo.token_id, // - token (hex str with 0x prefix < prime)
  feeInfo.source_vault_id, // - fee_source_vault_id (uint64)
  feeInfo.fee_limit // - amount (uint63 decimal str)
);

assert(
  msgHash ===
    testData.meta_data.transfer_order_with_fee.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.transfer_order.message_hash.substring(2)
);

// The following is the JSON representation of a transfer:
console.log('Transfer With Fee JSON representation: ');
console.log(transfer);
console.log('\n');
```

### StarkEx conditional Transfer with fee

```javascript
const privateKey =
  testData.meta_data.conditional_transfer_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) ===
    testData.conditional_transfer_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.conditional_transfer_order.public_key.substring(
          2
        )}`
);

const transfer = testData.conditional_transfer_order;
const feeInfo = testData.fee_info_user;
const msgHash = starkwareCrypto.getTransferMsgHashWithFee(
  transfer.amount, // - amount (uint63 decimal str)
  transfer.nonce, // - nonce (uint31)
  transfer.sender_vault_id, // - sender_vault_id (uint64)
  transfer.token, // - token (hex str with 0x prefix < prime)
  transfer.target_vault_id, // - target_vault_id (uint64)
  transfer.target_public_key, // - target_public_key (hex str with 0x prefix < prime)
  transfer.expiration_timestamp, // - expiration_timestamp (uint22)
  feeInfo.token_id, // - token (hex str with 0x prefix < prime)
  feeInfo.source_vault_id, // - fee_source_vault_id (uint64)
  feeInfo.fee_limit, // - amount (uint63 decimal str)
  transfer.condition // - condition (hex str with 0x prefix < prime)
);

assert(
  msgHash ===
    testData.meta_data.conditional_transfer_order_with_fee.message_hash.substring(
      2
    ),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.conditional_transfer_order.message_hash.substring(2)
);

// The following is the JSON representation of a transfer:
console.log('Conditional Transfer With Fee JSON representation: ');
console.log(transfer);
console.log('\n');
```

### Adding a matching order to create a settlement

```javascript
const starkwareCrypto = require('@starkware-libs/starkware-crypto-utils');
const testData = require('test/config/signature_test_data.json');

const privateKey = testData.meta_data.party_b_order.private_key.substring(2);
const keyPair = starkwareCrypto.ec.keyFromPrivate(privateKey, 'hex');
const publicKey = starkwareCrypto.ec.keyFromPublic(
  keyPair.getPublic(true, 'hex'),
  'hex'
);
const publicKeyX = publicKey.pub.getX();

assert(
  publicKeyX.toString(16) ===
    testData.settlement.party_b_order.public_key.substring(2),
  `Got: ${publicKeyX.toString(16)}.
        Expected: ${testData.settlement.party_b_order.public_key.substring(2)}`
);

const {party_b_order: partyBOrder} = testData.settlement;
const msgHash = starkwareCrypto.getLimitOrderMsgHash(
  partyBOrder.vault_id_sell, // - vault_sell (uint31)
  partyBOrder.vault_id_buy, // - vault_buy (uint31)
  partyBOrder.amount_sell, // - amount_sell (uint63 decimal str)
  partyBOrder.amount_buy, // - amount_buy (uint63 decimal str)
  partyBOrder.token_sell, // - token_sell (hex str with 0x prefix < prime)
  partyBOrder.token_buy, // - token_buy (hex str with 0x prefix < prime)
  partyBOrder.nonce, // - nonce (uint31)
  partyBOrder.expiration_timestamp // - expiration_timestamp (uint22)
);

assert(
  msgHash === testData.meta_data.party_b_order.message_hash.substring(2),
  `Got: ${msgHash}. Expected: ` +
    testData.meta_data.party_b_order.message_hash.substring(2)
);

const msgSignature = starkwareCrypto.sign(keyPair, msgHash);
const {r, s} = msgSignature;

assert(starkwareCrypto.verify(publicKey, msgHash, msgSignature));
assert(
  r.toString(16) === partyBOrder.signature.r.substring(2),
  `Got: ${r.toString(16)}. Expected: ${partyBOrder.signature.r.substring(2)}`
);
assert(
  s.toString(16) === partyBOrder.signature.s.substring(2),
  `Got: ${s.toString(16)}. Expected: ${partyBOrder.signature.s.substring(2)}`
);

// The following is the JSON representation of a settlement:
console.log('Settlement JSON representation: ');
console.log(testData.settlement);
```

## Valid transfer with sender_vault_id=2\*\*63+10

```javascript
const transfer = testData.transfer_order_2nd_valid_range;
const feeInfo = testData.fee_info_user;


CHANGELOG.md
### [0.2.1](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.2.0...v0.2.1) (2023-11-15)


### 🔧 Fixes

* require from module's root returns an empty object. ([#20](https://github.com/starkware-libs/starkware-crypto-utils/issues/20)) ([e141e88](https://github.com/starkware-libs/starkware-crypto-utils/commit/e141e886986864fbb9ac2b1cb69f07af05afe47d))


### 💉 Improvements

* optimize `pedersen` function ([#17](https://github.com/starkware-libs/starkware-crypto-utils/issues/17)) ([c73ee56](https://github.com/starkware-libs/starkware-crypto-utils/commit/c73ee565a865b88564441e21d64332478dca2455))

### [0.2.1-dev.2](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.2.1-dev.1...v0.2.1-dev.2) (2023-11-14)


### 💉 Improvements

* optimize `pedersen` function ([#17](https://github.com/starkware-libs/starkware-crypto-utils/issues/17)) ([c73ee56](https://github.com/starkware-libs/starkware-crypto-utils/commit/c73ee565a865b88564441e21d64332478dca2455))

### [0.2.1-dev.1](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.2.0...v0.2.1-dev.1) (2023-11-14)


### 🔧 Fixes

* require from module's root returns an empty object. ([#20](https://github.com/starkware-libs/starkware-crypto-utils/issues/20)) ([e141e88](https://github.com/starkware-libs/starkware-crypto-utils/commit/e141e886986864fbb9ac2b1cb69f07af05afe47d))

## [0.2.0](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.1.0...v0.2.0) (2023-08-14)


### 📚 Docs

* fixed URL to Pedersen hash documentation ([#15](https://github.com/starkware-libs/starkware-crypto-utils/issues/15)) ([23f09b4](https://github.com/starkware-libs/starkware-crypto-utils/commit/23f09b44369a58d2f9169291e103e33cb99ace52))


### 🔧 Fixes

* use ESM export from index file ([01e5c58](https://github.com/starkware-libs/starkware-crypto-utils/commit/01e5c58b09b46bac6c4b832b4a8a8601f6075db0))


### 🧩 Features

* polyfill node core modules ([#16](https://github.com/starkware-libs/starkware-crypto-utils/issues/16)) ([ee6290c](https://github.com/starkware-libs/starkware-crypto-utils/commit/ee6290cee5aaca92d24ce80e3c1d0f9e4ee736a6))

## [0.2.0-dev.1](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.1.1-dev.2...v0.2.0-dev.1) (2023-08-14)


### 🧩 Features

* polyfill node core modules ([#16](https://github.com/starkware-libs/starkware-crypto-utils/issues/16)) ([ee6290c](https://github.com/starkware-libs/starkware-crypto-utils/commit/ee6290cee5aaca92d24ce80e3c1d0f9e4ee736a6))

### [0.1.1-dev.2](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.1.1-dev.1...v0.1.1-dev.2) (2023-08-13)


### 🔧 Fixes

* use ESM export from index file ([01e5c58](https://github.com/starkware-libs/starkware-crypto-utils/commit/01e5c58b09b46bac6c4b832b4a8a8601f6075db0))

### [0.1.1-dev.1](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.1.0...v0.1.1-dev.1) (2023-05-15)


### 📚 Docs

* fixed URL to Pedersen hash documentation ([#15](https://github.com/starkware-libs/starkware-crypto-utils/issues/15)) ([23f09b4](https://github.com/starkware-libs/starkware-crypto-utils/commit/23f09b44369a58d2f9169291e103e33cb99ace52))

## [0.1.0](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.0.2...v0.1.0) (2023-05-07)


### 🧩 Features

* exporting getAssetSelector function ([#8](https://github.com/starkware-libs/starkware-crypto-utils/issues/8)) ([1a945b2](https://github.com/starkware-libs/starkware-crypto-utils/commit/1a945b2ffbdf56c68bb7606c45341c9d117422ad))
* typescript support ([#10](https://github.com/starkware-libs/starkware-crypto-utils/issues/10)) ([bcdb84b](https://github.com/starkware-libs/starkware-crypto-utils/commit/bcdb84b829897af809886e60a3fe284e1dde1eae))

## [0.1.0-dev.1](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.0.1...v0.1.0-dev.1) (2023-05-07)


### 🧩 Features

* exporting getAssetSelector function ([#8](https://github.com/starkware-libs/starkware-crypto-utils/issues/8)) ([1a945b2](https://github.com/starkware-libs/starkware-crypto-utils/commit/1a945b2ffbdf56c68bb7606c45341c9d117422ad))
* new APIs (getPrivateKeyFromEthSignature, privateToStarkKey) ([#3](https://github.com/starkware-libs/starkware-crypto-utils/issues/3)) ([bf4066c](https://github.com/starkware-libs/starkware-crypto-utils/commit/bf4066c81e4389daea0c388370ac509d89f1a3e1))
* typescript support ([#10](https://github.com/starkware-libs/starkware-crypto-utils/issues/10)) ([bcdb84b](https://github.com/starkware-libs/starkware-crypto-utils/commit/bcdb84b829897af809886e60a3fe284e1dde1eae))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.2](https://github.com/starkware-libs/starkware-crypto-utils/compare/v0.0.2-0...v0.0.2) (2022-03-31)

### 0.0.1 (2022-03-08)

### Features

- init project ([9759ffa](https://github.com/starkware-libs/starkware-crypto-utils/commit/9759ffac52538345524c90a784e653c95d8899f5))


src/js/signature.js
/////////////////////////////////////////////////////////////////////////////////
// Copyright 2019 StarkWare Industries Ltd.                                    //
//                                                                             //
// Licensed under the Apache License, Version 2.0 (the "License").             //
// You may not use this file except in compliance with the License.            //
// You may obtain a copy of the License at                                     //
//                                                                             //
// https://www.starkware.co/open-source-license/                               //
//                                                                             //
// Unless required by applicable law or agreed to in writing,                  //
// software distributed under the License is distributed on an "AS IS" BASIS,  //
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    //
// See the License for the specific language governing permissions             //
// and limitations under the License.                                          //
/////////////////////////////////////////////////////////////////////////////////

const BN = require('bn.js');
const hash = require('hash.js');
const {curves: eCurves, ec: EllipticCurve} = require('elliptic');
const assert = require('assert');
const {
  assertInRange,
  assertInMultiRange,
  Range
} = require('./message_utils.js');
const constantPointsHex = require('../config/constant_points.json');

// Equals 2**251 + 17 * 2**192 + 1.
const prime = new BN(
  '800000000000011000000000000000000000000000000000000000000000001',
  16
);
// Equals 2**251. This value limits msgHash and the signature parts.
const maxEcdsaVal = new BN(
  '800000000000000000000000000000000000000000000000000000000000000',
  16
);

// Generate BN of used constants.
const zeroBn = new BN('0', 16);
const oneBn = new BN('1', 16);
const twoBn = new BN('2', 16);
const threeBn = new BN('3', 16);
const fourBn = new BN('4', 16);
const fiveBn = new BN('5', 16);
const twoPow22Bn = new BN('400000', 16);
const twoPow31Bn = new BN('80000000', 16);
const twoPow63Bn = new BN('8000000000000000', 16);
const vaultIDAllowedRanges = [
  new Range(zeroBn, twoPow31Bn),
  new Range(twoPow63Bn, twoPow63Bn.add(twoPow31Bn))
];

// Create an EC curve with stark curve parameters.
const starkEc = new EllipticCurve(
  new eCurves.PresetCurve({
    type: 'short',
    prime: null,
    p: prime,
    a: '00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001',
    b: '06f21413 efbe40de 150e596d 72f7a8c5 609ad26c 15c915c1 f4cdfcb9 9cee9e89',
    n: '08000000 00000010 ffffffff ffffffff b781126d cae7b232 1e66a241 adc64d2f',
    hash: hash.sha256,
    gRed: false,
    g: constantPointsHex[1]
  })
);

const constantPoints = constantPointsHex.map(coords =>
  starkEc.curve.point(new BN(coords[0], 16), new BN(coords[1], 16))
);
const shiftPoint = constantPoints[0];

/*
  Checks that the string str start with '0x'.
*/
function hasHexPrefix(str) {
  return str.substring(0, 2) === '0x';
}

/*
 Full specification of the hash function can be found here:
   https://starkware.co/starkex/docs/signatures.html#pedersen-hash-function
 shiftPoint was added for technical reasons to make sure the zero point on the elliptic curve does
 not appear during the computation. constantPoints are multiples by powers of 2 of the constant
 points defined in the documentation.
*/
function pedersen(input) {
  let point = shiftPoint;
  for (let i = 0; i < input.length; i++) {
    let x = new BN(input[i], 16);
    // assert(x.gte(zeroBn) && x.lt(prime), 'Invalid input: ' + input[i]);
    for (let j = 0; j < 252; j++) {
      const pt = constantPoints[2 + i * 252 + j];
      assert(!point.getX().eq(pt.getX()));
      if (x.and(oneBn).toNumber() !== 0) {
        point = point.add(pt);
      }
      x = x.shrn(1);
    }
  }
  return point.getX().toString(16);
}

function hashMsg({
  instructionType,
  amountBn,
  zeroBn,
  nonceBn,
  feeBn,
  action,
  receiver,
  message,
}) {
  let packedMessage = instructionType;
  packedMessage = packedMessage.ushln(63).add(amountBn);
  if (feeBn && instructionType.eq(twoBn)) {
    packedMessage = packedMessage.ushln(63).add(feeBn);
  }
  packedMessage = packedMessage.ushln(31).add(nonceBn);
  let tmpHash
  if (receiver && instructionType.eq(oneBn)) {
    tmpHash = pedersen([
      pedersen([action, message]),
      receiver
    ]);
  } else {
    tmpHash = pedersen([action, message])
  }
  
  const msgHash = pedersen([tmpHash, packedMessage.toString(16)]);

  const msgHashBN = new BN(msgHash, 16);
  assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
  return msgHash;
}

/*
 Serializes the transfer message in the canonical format expected by the verifier.
 The sender transfer 'amount' coins of 'token' from vault with id senderVaultId to vault with id
 receiverVaultId. The receiver's public key is receiverPublicKey.
 If a condition is added, it is verified before executing the transfer. The format of the condition
 is defined by the application.
 Expected types:
 ---------------
 amount - uint63 (as decimal string)
 nonce - uint31 (as int)
 senderVaultId uint31 (as int)
 token - uint256 field element strictly less than the prime (as hex string with 0x)
 receiverVaultId - uint31 (as int)
 receiverPublicKey - uint256 field element strictly less than the prime (as hex string with 0x)
 expirationTimestamp - uint22 (as int).
 condition - uint256 field element strictly less than the prime (as hex string with 0x)
*/
function getTransferMsgHash({
  action, nonce, message
}) {
  const { receiver, amount, name } = action;
  assert(name === 'transfer', 'Invalid action name');
  assert(
      hasHexPrefix(receiver) && hasHexPrefix(message) && hasHexPrefix(name),
    'Hex strings expected to be prefixed with 0x.'
  );
  const amountBn = new BN(amount, 10);
  const nonceBn = new BN(nonce);
  const receiverBn = new BN(receiver.substring(2), 16);
  const actionBn = new BN(name.substring(2), 16);
  assertInRange(amountBn, zeroBn, twoPow63Bn);
  assertInRange(nonceBn, zeroBn, twoPow31Bn);
  assertInRange(actionBn, zeroBn, prime);
  assertInRange(receiverBn, zeroBn, prime);

  const instructionType = oneBn;
  
  return hashMsg({
    instructionType,
    amountBn,
    zeroBn,
    nonceBn,
    action: action.substring(2),
    receiver: receiver.substring(2),
    message: message.substring(2),
  });
}

function getWithdrawMsgHash({
  action, nonce, message
}) {
  const { fee, amount, name } = action;
  assert(name === 'withdraw', 'Invalid action name');
  assert(
      hasHexPrefix(message) && hasHexPrefix(name),
    'Hex strings expected to be prefixed with 0x.'
  );
  const amountBn = new BN(amount, 10);
  const feeBn = new BN(fee, 10);
  const nonceBn = new BN(nonce);
  const actionBn = new BN(name.substring(2), 16);
  assertInRange(amountBn, zeroBn, twoPow63Bn);
  assertInRange(feeBn, zeroBn, twoPow63Bn);
  assertInRange(nonceBn, zeroBn, twoPow31Bn);
  assertInRange(actionBn, zeroBn, prime);

  const instructionType = twoBn;
  
  return hashMsg({
    instructionType,
    amountBn,
    zeroBn,
    nonceBn,
    feeBn,
    action: action.substring(2),
    message: message.substring(2),
  });
}

/*
 Same as getTransferMsgHash, but also requires the fee info.

 Expected types of fee info params:
 ---------------
 feeVaultId - uint64 (as int)
 feeLimit - uint63 (as decimal string)
 feeToken - uint256 field element strictly less than the prime (as hex string with 0x)
*/
function getTransferMsgHashWithFee(
  amount,
  nonce,
  senderVaultId,
  token,
  receiverVaultId,
  receiverStarkKey,
  expirationTimestamp,
  feeToken,
  feeVaultId,
  feeLimit,
  condition = null
) {
  assert(
    hasHexPrefix(feeToken) &&
      hasHexPrefix(token) &&
      hasHexPrefix(receiverStarkKey) &&
      (condition === null || hasHexPrefix(condition)),
    'Hex strings expected to be prefixed with 0x.'
  );
  const amountBn = new BN(amount, 10);
  const nonceBn = new BN(nonce);
  const senderVaultIdBn = new BN(senderVaultId);
  const tokenBn = new BN(token.substring(2), 16);
  const receiverVaultIdBn = new BN(receiverVaultId);
  const receiverStarkKeyBn = new BN(receiverStarkKey.substring(2), 16);
  const expirationTimestampBn = new BN(expirationTimestamp);
  const feeTokenBn = new BN(feeToken);
  const feeVaultIdBn = new BN(feeVaultId);
  const feeLimitBn = new BN(feeLimit);

  assertInRange(amountBn, zeroBn, twoPow63Bn);
  assertInRange(nonceBn, zeroBn, twoPow31Bn);
  assertInMultiRange(senderVaultIdBn, vaultIDAllowedRanges);
  assertInRange(tokenBn, zeroBn, prime);
  assertInMultiRange(receiverVaultIdBn, vaultIDAllowedRanges);
  assertInRange(receiverStarkKeyBn, zeroBn, prime);
  assertInRange(expirationTimestampBn, zeroBn, twoPow22Bn);
  assertInRange(feeTokenBn, zeroBn, prime);
  assertInMultiRange(feeVaultIdBn, vaultIDAllowedRanges);
  assertInRange(feeLimitBn, zeroBn, twoPow63Bn);

  let instructionType = fourBn;
  if (condition !== null) {
    condition = condition.substring(2);
    assertInRange(new BN(condition), zeroBn, prime, 'condition');
    instructionType = fiveBn;
  }
  return hashTransferMsgWithFee(
    instructionType,
    senderVaultIdBn,
    receiverVaultIdBn,
    amountBn,
    nonceBn,
    expirationTimestampBn,
    token.substring(2),
    receiverStarkKey.substring(2),
    feeToken.substring(2),
    feeVaultIdBn,
    feeLimitBn,
    condition
  );
}

/*
 The function _truncateToN in lib/elliptic/ec/index.js does a shift-right of delta bits,
 if delta is positive, where
   delta = msgHash.byteLength() * 8 - starkEx.n.bitLength().
 This function does the opposite operation so that
   _truncateToN(fixMsgHashLen(msgHash)) == msgHash.
*/
function fixMsgHashLen(msgHash) {
  // Convert to BN to remove leading zeros.
  msgHash = new BN(msgHash, 16).toString(16);

  if (msgHash.length <= 62) {
    // In this case, msgHash should not be transformed, as the byteLength() is at most 31,
    // so delta < 0 (see _truncateToN).
    return msgHash;
  }
  assert(msgHash.length === 63);
  // In this case delta will be 4 so we perform a shift-left of 4 bits by adding a zero.
  return msgHash + '0';
}

/*
 Signs a message using the provided key.
 privateKey should be an elliptic.keyPair with a valid private key.
 Returns an elliptic.Signature.
*/
function sign(privateKey, msgHash) {
  const msgHashBN = new BN(msgHash, 16);
  // Verify message hash has valid length.
  assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
  const msgSignature = privateKey.sign(fixMsgHashLen(msgHash));
  const {r, s} = msgSignature;
  const w = s.invm(starkEc.n);
  // Verify signature has valid length.
  assertInRange(r, oneBn, maxEcdsaVal, 'r');
  assertInRange(s, oneBn, starkEc.n, 's');
  assertInRange(w, oneBn, maxEcdsaVal, 'w');
  return msgSignature;
}

/*
 Verifies a message using the provided key.
 publicKey should be an elliptic.keyPair with a valid public key.
 msgSignature should be an elliptic.Signature.
 Returns a boolean true if the verification succeeds.
*/
function verify(publicKey, msgHash, msgSignature) {
  const msgHashBN = new BN(msgHash, 16);
  // Verify message hash has valid length.
  assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
  const {r, s} = msgSignature;
  const w = s.invm(starkEc.n);
  // Verify signature has valid length.
  assertInRange(r, oneBn, maxEcdsaVal, 'r');
  assertInRange(s, oneBn, starkEc.n, 's');
  assertInRange(w, oneBn, maxEcdsaVal, 'w');
  return publicKey.verify(fixMsgHashLen(msgHash), msgSignature);
}

module.exports = {
  prime,
  ec: starkEc,
  constantPoints,
  shiftPoint,
  maxEcdsaVal, // Data.
  Range, // Class.
  pedersen,
  getTransferMsgHash,
  getWithdrawMsgHash,
  sign,
  verify,
  assertInRange,
  assertInMultiRange,
  getTransferMsgHashWithFee,
};
